
# Muhammad Irvan Hakim
# 2109106057
# Posttest 4 (Data Preprocessing)

#### Import Library pandas

In [216]:
import pandas as pd

#### Menginput File CSV Menggunakan Pandas

In [217]:
df = pd.read_csv('Wine_Quality_Dataset_1(Original).csv')
df

Unnamed: 0  fixed_acidity  volatile_acidity  citric_acid  \
0              0            7.4              0.70         0.00   
1              1            7.8              0.88         0.00   
2              2            7.8              0.76         0.04   
3              3           11.2              0.28         0.56   
4              4            7.4              0.66         0.00   
...          ...            ...               ...          ...   
5315        5315            6.2              0.21         0.29   
5316        5316            6.6              0.32         0.36   
5317        5317            6.5              0.24         0.19   
5318        5318            5.5              0.29         0.30   
5319        5319            6.0              0.21         0.38   

      residual_sugar  chlorides  free_sulfur_dioxide  total_sulfur_dioxide  \
0                1.9      0.076                 11.0                  34.0   
1                2.6      0.098                 25.0                  67.0   
2                2.3      0.092                 15.0                  54.0   
3                1.9      0.075                 17.0                  60.0   
4                1.8      0.075                 13.0                  40.0   
...              ...        ...                  ...                   ...   
5315             1.6      0.039                 24.0                  92.0   
5316             8.0      0.047                 57.0                 168.0   
5317             1.2      0.041                 30.0                 111.0   
5318             1.1      0.022                 20.0                 110.0   
5319             0.8      0.020                 22.0                  98.0   

      density    pH  sulphates  alcohol  quality  is_red  
0     0.99780  3.51       0.56      9.4        5       1  
1     0.99680  3.20       0.68      9.8        5       1  
2     0.99700  3.26       0.65      9.8        5       1  
3     0.99800  3.16       0.58      9.8        6       1  
4     0.99780  3.51       0.56      9.4        5       1  
...       ...   ...        ...      ...      ...     ...  
5315  0.99114  3.27       0.50     11.2        6       0  
5316  0.99490  3.15       0.46      9.6        5       0  
5317  0.99254  2.99       0.46      9.4        6       0  
5318  0.98869  3.34       0.38     12.8        7       0  
5319  0.98941  3.26       0.32     11.8        6       0  

[5320 rows x 14 columns]

## 1. Bagi dataset menjadi training set dan testing set dengan proporsi 70:30

#### Import train_test_split dari Library Sklearn

In [218]:
from sklearn.model_selection import train_test_split

#### Deklarasi variable X dan variable y

In [219]:
# Subset feature dan target
X = df.iloc[:,:-1] # Target
y = df.iloc[:,-1] # Feature
X,y

(      Unnamed: 0  fixed_acidity  volatile_acidity  citric_acid  \
 0              0            7.4              0.70         0.00   
 1              1            7.8              0.88         0.00   
 2              2            7.8              0.76         0.04   
 3              3           11.2              0.28         0.56   
 4              4            7.4              0.66         0.00   
 ...          ...            ...               ...          ...   
 5315        5315            6.2              0.21         0.29   
 5316        5316            6.6              0.32         0.36   
 5317        5317            6.5              0.24         0.19   
 5318        5318            5.5              0.29         0.30   
 5319        5319            6.0              0.21         0.38   
 
       residual_sugar  chlorides  free_sulfur_dioxide  total_sulfur_dioxide  \
 0                1.9      0.076                 11.0                  34.0   
 1                2.6      0.098    

#### Buat kelas train_test_split. Nilai test_size kita kasih 0.3 saja agar 30% dari data akan menjadi testing set dan 70% dari data akan menjadi training set

In [220]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

#### Tampilkan jumlah baris dan kolom dari train set dan test set

In [221]:
print("Dimensi X_train : ", X_train.shape)
print("Dimensi X_train : ", X_test.shape)
print("Dimensi y_train : ", y_train.shape)
print("Dimensi y_test : ", y_test.shape)

Dimensi X_train :  (3724, 13)
Dimensi X_train :  (1596, 13)
Dimensi y_train :  (3724,)
Dimensi y_test :  (1596,)


## 2. Lakukan normalisasi data pada salah satu attribute menggunakan Min Max scaler (buatlah copy dataset terlebih dahulu)

#### Buat Copy Dataset

In [222]:
df2 = pd.read_csv('Wine_Quality_Dataset_2(Normalisasi).csv')

#### Jalankan Kodingan Di Bawah Ini Untuk Melakukan Normalisasi Data Pada Atribut 'total_sulfur_dioxide'

In [225]:
# Import MinMaxScaler Dari Library sklearn
from sklearn.preprocessing import MinMaxScaler

# Membuat objek MinMaxScaler
mms = MinMaxScaler()

# Tampilkan nilai dari atribut 'total_sulfur_dioxide' sebelum di scaling Normalisasi
print("Nilai Dari Atribut 'total_sulfur_dioxide' Sebelum Scaling Normalisasi : ")
print(df2['total_sulfur_dioxide'])

# Normalisasi atribut 'total_sulfur_dioxide'
scaled_data = mms.fit_transform(df2[['total_sulfur_dioxide']])

# Tampilkan atribut 'total_sulfur_dioxide' setelah di scaling Normalisasi
print("\nNilai Dari Atribut 'total_sulfur_dioxide' Setelah Scaling Normalisasi : ")
print(scaled_data)

Nilai Dari Atribut 'total_sulfur_dioxide' Sebelum Scaling Normalisasi : 
0        34.0
1        67.0
2        54.0
3        60.0
4        40.0
        ...  
5315     92.0
5316    168.0
5317    111.0
5318    110.0
5319     98.0
Name: total_sulfur_dioxide, Length: 5320, dtype: float64

Nilai Dari Atribut 'total_sulfur_dioxide' Setelah Scaling Normalisasi : 
[[0.06451613]
 [0.140553  ]
 [0.11059908]
 ...
 [0.24193548]
 [0.23963134]
 [0.21198157]]


#### Penjelasan: Isi nilai dari atribut 'total_sulfur_dioxide' yang awalnya tidak memiliki batas nilai minimal dan batas nilai maksimal, sekarang memiliki batas nilai minimal berupa 0 dan batas nilai maksimal berupa 1

## 3. Lakukan standarisasi pada dataset (buatlah copy dataset terlebih dahulu)

#### Buat Copy Dataset

In [226]:
df3 = pd.read_csv('Wine_Quality_Dataset_3(Standarisasi).csv')

#### Jalankan Kodingan Di Bawah Ini Untuk Melakukan Standarisasi Dataset

In [227]:
# Import StandardScaler dari library sklearn
from sklearn.preprocessing import StandardScaler

# Import Numpy
import numpy as np

# Ubah dataset menjadi array agar bisa dihitung nilai standar deviasi nya
data = df3.to_numpy()

# Membuat objek StandardScaler
ss = StandardScaler()

# Standarisasi dataset
scaled_data = ss.fit_transform(data)

# Tampilkan nilai dataset sebelum di scaling Standarisasi
print("Nilai Data Sebelum Scaling Standarisasi : ")
print(data)
print("Nilai standar deviasi: ", np.std(data))

# Tampilkan nilai dataset setelah di scaling Standarisasi
print("\nNilai Data Setelah Scaling Standarisasi : ")
print(scaled_data)
print("Nilai standar deviasi: ", np.std(scaled_data))

Nilai Data Sebelum Scaling Standarisasi : 
[[0.000e+00 7.400e+00 7.000e-01 ... 9.400e+00 5.000e+00 1.000e+00]
 [1.000e+00 7.800e+00 8.800e-01 ... 9.800e+00 5.000e+00 1.000e+00]
 [2.000e+00 7.800e+00 7.600e-01 ... 9.800e+00 5.000e+00 1.000e+00]
 ...
 [5.317e+03 6.500e+00 2.400e-01 ... 9.400e+00 6.000e+00 0.000e+00]
 [5.318e+03 5.500e+00 2.900e-01 ... 1.280e+01 7.000e+00 0.000e+00]
 [5.319e+03 6.000e+00 2.100e-01 ... 1.180e+01 6.000e+00 0.000e+00]]
Nilai standar deviasi:  796.1469447113785

Nilai Data Setelah Scaling Standarisasi : 
[[-1.73172526  0.14006434  2.11534878 ... -0.96915199 -0.90449728
   1.70723259]
 [-1.73107412  0.44319875  3.18529696 ... -0.63183308 -0.90449728
   1.70723259]
 [-1.73042297  0.44319875  2.47199817 ... -0.63183308 -0.90449728
   1.70723259]
 ...
 [ 1.73042297 -0.54198809 -0.61896323 ... -0.96915199  0.23226755
  -0.58574327]
 [ 1.73107412 -1.29982413 -0.32175541 ...  1.89805872  1.36903238
  -0.58574327]
 [ 1.73172526 -0.92090611 -0.79728793 ...  1.05476146

#### Penjelasan: Pada dataset tersebut, nilai standar deviasi sebelum melakukan standarisasi adalah 796.1469447113785. Setelah melakukan standarisasi, nilai standar deviasi nya berubah menjadi 1.

## 4. Lakukan Data cleaning pada data dengan nilai null (jika tidak ada nilai null pada dataset, maka buatlah menjadi ada). Ganti nilai null sesuai ketentuan. (bilangan bulat : median/modus, bilangan desimal : mean, tulisan : modus). Nilai plus jika strategy yang digunakan menggunakan 3 nilai

#### Import SimpleImputer Yang Akan Digunakan Untuk Mengganti Nilai Null

In [234]:
from sklearn.impute import SimpleImputer

#### Buat Copy Dataset Yang Ada Nilai Null Di Dalamnya

In [235]:
df4 = pd.read_csv('Wine_Quality_Dataset_4(Yang_ada_Null_nya).csv')

#### Periksa jumlah data pada dataset yang memiliki nilai null

In [236]:
print("Jumlah record yang memiliki nilai null:")
print(df4.isna().sum())

Jumlah record yang memiliki nilai null:
Unnamed: 0               0
fixed_acidity            0
volatile_acidity         0
citric_acid              0
residual_sugar           0
chlorides                0
free_sulfur_dioxide      0
total_sulfur_dioxide     0
density                  0
pH                      25
sulphates                0
alcohol                  0
quality                 13
is_red                  10
dtype: int64


#### Periksa Dataset sebelum memulai proses data cleaning

In [237]:
print("Sebelum Data Cleaning Pada Nilai Null : ")
df4

Sebelum Data Cleaning Pada Nilai Null : 


Unnamed: 0  fixed_acidity  volatile_acidity  citric_acid  \
0              0            7.4              0.70         0.00   
1              1            7.8              0.88         0.00   
2              2            7.8              0.76         0.04   
3              3           11.2              0.28         0.56   
4              4            7.4              0.66         0.00   
...          ...            ...               ...          ...   
5315        5315            6.2              0.21         0.29   
5316        5316            6.6              0.32         0.36   
5317        5317            6.5              0.24         0.19   
5318        5318            5.5              0.29         0.30   
5319        5319            6.0              0.21         0.38   

      residual_sugar  chlorides  free_sulfur_dioxide  total_sulfur_dioxide  \
0                1.9      0.076                 11.0                  34.0   
1                2.6      0.098                 25.0                  67.0   
2                2.3      0.092                 15.0                  54.0   
3                1.9      0.075                 17.0                  60.0   
4                1.8      0.075                 13.0                  40.0   
...              ...        ...                  ...                   ...   
5315             1.6      0.039                 24.0                  92.0   
5316             8.0      0.047                 57.0                 168.0   
5317             1.2      0.041                 30.0                 111.0   
5318             1.1      0.022                 20.0                 110.0   
5319             0.8      0.020                 22.0                  98.0   

      density    pH  sulphates  alcohol  quality  is_red  
0     0.99780   NaN       0.56      9.4      NaN     NaN  
1     0.99680   NaN       0.68      9.8      NaN     NaN  
2     0.99700   NaN       0.65      9.8      NaN     NaN  
3     0.99800   NaN       0.58      9.8      NaN     NaN  
4     0.99780  3.31       0.56      9.4      8.0     1.0  
...       ...   ...        ...      ...      ...     ...  
5315  0.99114  3.27       0.50     11.2      6.0     0.0  
5316  0.99490  3.15       0.46      9.6      5.0     0.0  
5317  0.99254  2.99       0.46      9.4      6.0     0.0  
5318  0.98869  3.34       0.38     12.8      7.0     0.0  
5319  0.98941  3.26       0.32     11.8      6.0     0.0  

[5320 rows x 14 columns]

#### Jalankan kodingan dibawah ini untuk melakukan Data Cleaning pada atribut pH, quality, dan is_red

In [238]:
# Mengganti nilai pH (bilangan desimal) yang berisi null menjadi bilangan mean
imputer = SimpleImputer(strategy='mean')
df4["pH"] = imputer.fit_transform(df4[["pH"]])

# Mengganti nilai quality (bilangan bulat) yang berisi null menjadi bilangan median
imputer = SimpleImputer(strategy='median')
df4["quality"] = imputer.fit_transform(df4[["quality"]])

# Mengganti nilai is_red (bilangan bulat) yang berisi null menjadi bilangan modus
imputer = SimpleImputer(strategy='most_frequent')
df4["is_red"] = imputer.fit_transform(df4[["is_red"]])

#### Menampilkan Dataset setelah melakukan data cleaning

In [240]:
print("Setelah Data Cleaning Pada Nilai Null : ")
df4

Setelah Data Cleaning Pada Nilai Null : 


Unnamed: 0  fixed_acidity  volatile_acidity  citric_acid  \
0              0            7.4              0.70         0.00   
1              1            7.8              0.88         0.00   
2              2            7.8              0.76         0.04   
3              3           11.2              0.28         0.56   
4              4            7.4              0.66         0.00   
...          ...            ...               ...          ...   
5315        5315            6.2              0.21         0.29   
5316        5316            6.6              0.32         0.36   
5317        5317            6.5              0.24         0.19   
5318        5318            5.5              0.29         0.30   
5319        5319            6.0              0.21         0.38   

      residual_sugar  chlorides  free_sulfur_dioxide  total_sulfur_dioxide  \
0                1.9      0.076                 11.0                  34.0   
1                2.6      0.098                 25.0                  67.0   
2                2.3      0.092                 15.0                  54.0   
3                1.9      0.075                 17.0                  60.0   
4                1.8      0.075                 13.0                  40.0   
...              ...        ...                  ...                   ...   
5315             1.6      0.039                 24.0                  92.0   
5316             8.0      0.047                 57.0                 168.0   
5317             1.2      0.041                 30.0                 111.0   
5318             1.1      0.022                 20.0                 110.0   
5319             0.8      0.020                 22.0                  98.0   

      density       pH  sulphates  alcohol  quality  is_red  
0     0.99780  3.22345       0.56      9.4      6.0     0.0  
1     0.99680  3.22345       0.68      9.8      6.0     0.0  
2     0.99700  3.22345       0.65      9.8      6.0     0.0  
3     0.99800  3.22345       0.58      9.8      6.0     0.0  
4     0.99780  3.31000       0.56      9.4      8.0     1.0  
...       ...      ...        ...      ...      ...     ...  
5315  0.99114  3.27000       0.50     11.2      6.0     0.0  
5316  0.99490  3.15000       0.46      9.6      5.0     0.0  
5317  0.99254  2.99000       0.46      9.4      6.0     0.0  
5318  0.98869  3.34000       0.38     12.8      7.0     0.0  
5319  0.98941  3.26000       0.32     11.8      6.0     0.0  

[5320 rows x 14 columns]

#### Periksa jumlah data pada dataset yang memiliki nilai null

In [166]:
print("Jumlah record yang memiliki nilai null:")
print(df4.isna().sum())

Jumlah record yang memiliki nilai null:
Unnamed: 0              0
fixed_acidity           0
volatile_acidity        0
citric_acid             0
residual_sugar          0
chlorides               0
free_sulfur_dioxide     0
total_sulfur_dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
is_red                  0
dtype: int64


#### Penjelasan: Dataset awalnya memiliki nilai NULL pada atribut pH, alcohol, dan is_red. Setelah melakukan data cleaning, nilai NULL pada atribut pH berubah menjadi bilangan mean, nilai NULL pada atribut quality berubah menjadi bilangan median, dan nilai NULL pada atribut is_red berubah menjadi bilangan modus. Dengan begini, dataset tidak lagi memiliki nilai NULL di dalam nya.    

## 5. Lakukan Data cleaning pada data dengan nilai duplikat. (Jika tidak ada nilai duplikat pada dataset, maka buatlah menjadi ada)

#### Buat Copy Dataset

In [193]:
df5 = pd.read_csv('Wine_Quality_Dataset_1(Original).csv')

#### Buat menambahkan nilai duplikat ke dalam dataset nya dikarenakan dataset saya tidak memiliki nilai duplikat

In [194]:
df5.iloc[1000, :] = df5.iloc[1001, :]
df5.iloc[2000, :] = df5.iloc[2001, :]
df5.iloc[3000, :] = df5.iloc[3001, :]
df5.iloc[4000, :] = df5.iloc[4001, :]

#### Buat mengecek nilai duplikat

In [195]:
print("Nilai Duplikat: ")
df5.duplicated().sum()

Nilai Duplikat: 


4

#### Buat menghapus nilai duplikat

In [196]:
df5.drop_duplicates(inplace=True)

#### Buat mengecek nilai duplikat

In [197]:
print("Nilai Duplikat: ")
df5.duplicated().sum()

Nilai Duplikat: 


0

#### Penjelasan: Dataset awalnya memiliki nilai duplikat di dalamnya. Setelah melakukan data cleaning, nilai duplikat yang ada didalam dataset langsung terhapus

## 6. Ganti tipe data salah satu attribute angka

#### Buat Copy Dataset

In [251]:
df6 = pd.read_csv('Wine_Quality_Dataset_1(Original).csv')

#### Periksa salah satu atribut dalam dataset yang bertipe angka menggunakan fungsi .dtypes

In [253]:
print("Tipe data kolom 'is_red' sebelum diganti : ")
df6['is_red'].dtypes

Tipe data kolom 'is_red' sebelum diganti : 


dtype('int64')

#### Buat mengganti tipe data atribut 'is_red' dari integer menjadi object

In [256]:
df6['is_red'] = df6['is_red'].astype(object)

#### Periksa tipe data atribut is_red

In [257]:
print("Tipe data kolom 'is_red' setelah diganti : ")
df6['is_red'].dtypes

Tipe data kolom 'is_red' setelah diganti : 


dtype('O')

#### Periksa informasi dari dataset untuk memastikan lagi bahwa tipe data atribut is_red sudah berubah menjadi object 

In [255]:
df6.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5320 entries, 0 to 5319
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            5320 non-null   int64  
 1   fixed_acidity         5320 non-null   float64
 2   volatile_acidity      5320 non-null   float64
 3   citric_acid           5320 non-null   float64
 4   residual_sugar        5320 non-null   float64
 5   chlorides             5320 non-null   float64
 6   free_sulfur_dioxide   5320 non-null   float64
 7   total_sulfur_dioxide  5320 non-null   float64
 8   density               5320 non-null   float64
 9   pH                    5320 non-null   float64
 10  sulphates             5320 non-null   float64
 11  alcohol               5320 non-null   float64
 12  quality               5320 non-null   int64  
 13  is_red                5320 non-null   object 
dtypes: float64(11), int64(2), object(1)
memory usage: 582.0+ KB


#### Penjelasan: atribut is_red yang awalnya bertipe integer sudah berubah menjadi bertipe object dengan cara menggunakan fungsi .astype(object)

## 7. Lakukan one hot encoding pada dataset yang kalian miliki

#### Import OneHotEncoder dari Library sklearn

In [203]:
from sklearn.preprocessing import OneHotEncoder

#### Buat copy dataset

In [259]:
df7 = pd.read_csv('Wine_Quality_Dataset_1(Original).csv')

#### Jalankan kodingan dibawah ini untuk melakukan one hot encoding

In [263]:
OHEncoder = OneHotEncoder(sparse_output=False)

# Memilih atribut is_red sebagai atribut yang akan dilakukan one hot encoding
is_redEnc = OHEncoder.fit_transform(df7[["is_red"]])
is_redEnc = pd.DataFrame(is_redEnc)

# Membuat atribut baru bernama "Anggur Biasa" yang mewakili nilai 0
df7["Anggur Biasa"] = is_redEnc[0]

# Membuat atribut baru bernama "Anggur Merah" yang mewakili nilai 1
df7["Anggur Merah"] = is_redEnc[1]

#### Tampilkan dataset setelah dilakukan nya One Hot Encoding

In [262]:
df7

Unnamed: 0  fixed_acidity  volatile_acidity  citric_acid  \
0              0            7.4              0.70         0.00   
1              1            7.8              0.88         0.00   
2              2            7.8              0.76         0.04   
3              3           11.2              0.28         0.56   
4              4            7.4              0.66         0.00   
...          ...            ...               ...          ...   
5315        5315            6.2              0.21         0.29   
5316        5316            6.6              0.32         0.36   
5317        5317            6.5              0.24         0.19   
5318        5318            5.5              0.29         0.30   
5319        5319            6.0              0.21         0.38   

      residual_sugar  chlorides  free_sulfur_dioxide  total_sulfur_dioxide  \
0                1.9      0.076                 11.0                  34.0   
1                2.6      0.098                 25.0                  67.0   
2                2.3      0.092                 15.0                  54.0   
3                1.9      0.075                 17.0                  60.0   
4                1.8      0.075                 13.0                  40.0   
...              ...        ...                  ...                   ...   
5315             1.6      0.039                 24.0                  92.0   
5316             8.0      0.047                 57.0                 168.0   
5317             1.2      0.041                 30.0                 111.0   
5318             1.1      0.022                 20.0                 110.0   
5319             0.8      0.020                 22.0                  98.0   

      density    pH  sulphates  alcohol  quality  is_red  Anggur Biasa  \
0     0.99780  3.51       0.56      9.4        5       1           0.0   
1     0.99680  3.20       0.68      9.8        5       1           0.0   
2     0.99700  3.26       0.65      9.8        5       1           0.0   
3     0.99800  3.16       0.58      9.8        6       1           0.0   
4     0.99780  3.51       0.56      9.4        5       1           0.0   
...       ...   ...        ...      ...      ...     ...           ...   
5315  0.99114  3.27       0.50     11.2        6       0           1.0   
5316  0.99490  3.15       0.46      9.6        5       0           1.0   
5317  0.99254  2.99       0.46      9.4        6       0           1.0   
5318  0.98869  3.34       0.38     12.8        7       0           1.0   
5319  0.98941  3.26       0.32     11.8        6       0           1.0   

      Anggur Merah  
0              1.0  
1              1.0  
2              1.0  
3              1.0  
4              1.0  
...            ...  
5315           0.0  
5316           0.0  
5317           0.0  
5318           0.0  
5319           0.0  

[5320 rows x 16 columns]

#### Penjelasan: Setelah melakukan one hot encoding pada dataset tersebut, dataset akan menambahkan 2 atribut baru, yaitu atribut Anggur Biasa dan Anggur Merah. Atribut Anggur Biasa bernilai 1 menandakan bahwa anggur tersebut tidak berwarna merah dan atribut Anggur Merah bernilai 1 menandakan bahwa anggur tersebut berwarna merah.